service-registry/registry-zero-config/src/main/java/org/apache/servicecomb/zeroconfig/client/ZeroConfigClient.java

```diff
@@ -0,0 +1,309 @@
+    FindInstancesResponse response = new FindInstancesResponse();
+    List<MicroserviceInstance> resultInstanceList = new ArrayList<>();
+
+    // 1.  find matched appId and serviceName from "Server"
+    List<ServerMicroserviceInstance> tempServerInstanceList = this.zeroConfigRegistryService.
+        findServiceInstances(appId, providerServiceName);
+
+    // 2.  find matched instance based on the strVersionRule
+    VersionRule versionRule = VersionRuleUtils.getOrCreate(strVersionRule);
+
+    ServerMicroserviceInstance latestVersionInstance = findLatestVersionInstance(
+        tempServerInstanceList, versionRule);
+    if (latestVersionInstance != null) {
+      Version latestVersion = VersionUtils.getOrCreate(latestVersionInstance.getVersion());
+      for (ServerMicroserviceInstance serverInstance : tempServerInstanceList) {
+        Version version = VersionUtils.getOrCreate(serverInstance.getVersion());
+        if (!versionRule.isMatch(version, latestVersion)) {
+          continue;
+        }
+        resultInstanceList.add(ClientUtil.convertToClientMicroserviceInstance(serverInstance));
+      }
+    }
+
+    response.setInstances(resultInstanceList);
+    resultMicroserviceInstances.setInstancesResponse(response);
+    return resultMicroserviceInstances;
+  }
+
+  private ServerMicroserviceInstance findLatestVersionInstance(
+      List<ServerMicroserviceInstance> instanceList, VersionRule versionRule) {
+    Version latestVersion = null;
+    ServerMicroserviceInstance latestVersionInstance = null;
+    for (ServerMicroserviceInstance serverInstance : instanceList) {
+      Version version = VersionUtils.getOrCreate(serverInstance.getVersion());
+      if (!versionRule.isAccept(version)) {
+        continue;
+      }
+
+      if (latestVersion == null || version.compareTo(latestVersion) > 0) {
+        latestVersion = version;
+        latestVersionInstance = serverInstance;
+      }
+    }
+    return latestVersionInstance;
+  }
+
+  private Map<String, String> prepareRegisterData() {
+    // set serviceId
+    Microservice selfMicroservice = ZeroConfigRegistration.INSTANCE.getSelfMicroservice();
+    MicroserviceInstance selfMicroserviceInstance = ZeroConfigRegistration.INSTANCE
+        .getSelfMicroserviceInstance();
+
+    if (selfMicroservice == null || selfMicroserviceInstance == null) {
+      return null;
+    }
+
+    String serviceId = selfMicroservice.getServiceId();
+    if (StringUtils.isEmpty(serviceId)) {
+      serviceId = ClientUtil.generateServiceId(selfMicroservice);
+      selfMicroservice.setServiceId(serviceId);
+      selfMicroserviceInstance.setServiceId(serviceId);
+    }
+
+    // set instanceId
+    String instanceId = selfMicroserviceInstance
+        .getInstanceId(); // allow client to set the instanceId
+    if (StringUtils.isEmpty(instanceId)) {
+      instanceId = ClientUtil.generateServiceInstanceId();
+      selfMicroserviceInstance.setInstanceId(instanceId);
+    }
```

Comment:
I think it is better to initalize instance id and service id in `init`.   And can avoid many `null` and `empty` checks here.

Reply:
ok